### ISSUE TITLE
Add option to make "cobalt publish" add date prefix to filename

### ISSUE THREAD

**Issue by Geobert** [cobalt-org/cobalt.rs#562]
To keep files in order by chronology, it would be cool to add the `YYYY-MM-DD-` prefix to the filename on `cobalt publish`

# Proposal

Activation by a configuration field in `_cobalt.yml`:
```yml
posts:
  publish_date_in_filename: true
```
Default:
- publish_date_in_filename true for posts
- publish_date_in_filename false for pages

`cobalt publish` behavior:
- Add or update date prefix on file name
- Set or update published_date in frontmatter

## Impact on inferring `published_date` from filename

For now, we will unconditionally still infer the date from the filename.
- Not everyone will be using `cobalt publish`
- Until we have Collections (#368), we sometimes have to imitate it.

An open question for later is whether we should make the inferring of the date exclusive to this flag being enabled.  That question is irrelevant until we have Collections support.

**Comment by epage** [comment]
Some interesting considerations
- This won't be applicable to all users
- This won't be applicable to all collections

Do we solve this with a command line flag, a global config option, a per-collection config option?

**Comment by epage** [comment]
> Which users?

Any users which want to not have the date in their filename

**Comment by Geobert** [comment]
So it's ok to have flag/option to activate it, isn't it?

**Comment by epage** [comment]
Possibly.  One of my concerns is providing too  many options, overwhelming people.  This is one of the reaosns I merged `cobalt serve` and `cobalt watch`.

**Comment by Geobert** [comment]
If the option are well documented, I don't see a problem, especially in this simple case. Or to simplify this, let's stick to global option only? No CLI flag.

**Comment by Geobert** [comment]
added proposal to the first post

**Comment by epage** [comment]
```yml
posts:
  add_date_prefix: true
```

We need to consider what name makes the intent clear.  This name doesn't help us know it is scoped to `cobalt publish` or that it is the `published_date`.

We can also look at re-framing the problem.  We could instead consider a flag to enable extracting the published_date from the filename and if it is enabled, then `cobalt publish` also prepends it.

**Comment by Geobert** [comment]
> We need to consider what name makes the intent clear.

ok, what about `prepend_date_to_file_name_on_publish`?

> a flag to enable extracting the published_date from the filename and if it is enabled, then cobalt publish also prepends it.

uh? I don't get it: extracting from the filename to add it again to the filename?

**Comment by Geobert** [comment]
According to the doc, we have:
> Recognized name formats include:
>    YYYY-MM-DD-slug.extension
 >   YYYY-MM-DD slug.extension
>    slug.extension

but I can't find the code that takes care of that Oo

**Comment by epage** [comment]
> uh? I don't get it: extracting from the filename to add it again to the filename?

Sorry  I wasn't clear.  One possible way of handling this is to put *all* date-in-name behavior behind the same flag. If you enable this flag
- We extract the published_date from the filename and put it in the frontmatter
- `cobalt publish` will prepend it to the filename

> but I can't find the code that takes care of that Oo

https://github.com/cobalt-org/cobalt.rs/blob/master/src/cobalt_model/frontmatter.rs#L373

**Comment by Geobert** [comment]
Thanks for the code!

So you're suggesting that we turn off the extract from filename by default? Won't this break current behaviour?

**Comment by Geobert** [comment]
I have a first implementation https://github.com/Geobert/cobalt.rs/tree/publish-add-date-prefix. I'll create the PR once we agree on the flag's name and behaviour and changed my code accordingly :)

**Comment by epage** [comment]
> So you're suggesting that we turn off the extract from filename by default? Won't this break current behaviour?

Think of my comment more as brainstorming than suggesting. Sometimes it is useful to not just attack a problem directly but look at alternative abstractions / mental models, particularly with how something interacts with other features. Sometimes it provides a better solution; sometimes it provides a refinement on the existing solution.

Yes, this would break compat. We could address this by making it the default. I do not consider changing `publish` or `new`s behavior is breaking.

**Comment by epage** [comment]
Eh, I'd say go ahead and create a PR.  The overhead is low if you already created a branch and it allows both discussions to happen in parallel.

**Comment by epage** [comment]
Even proposing is stronger than I'd like.  Both convey that I am in favor of the idea when I am neither in favor or against it.

The other thing about "suggesting" and american culture is it is very easy to have a "yes man" culture where when the "boss" speaks, people run with it.  There are whole discussions about leaders holding their ideas until the end so everyone feels safe bringing up their idea.

**Comment by Geobert** [comment]
I see what you mean :) Is there a word for "proposing without any judgement on the idea" in English?

What do you think of the "proposal" I made?

**Comment by epage** [comment]
Could you update the original issue with the proposal so we have it all summarized in one place rather than throughout this thread?

**Comment by Geobert** [comment]
done, with a new flag name as well: `publish_date_in_filename`

**Comment by epage** [comment]
I tweaked it some to try to make it more explicit.  Feel free to update it if I misunderstood

> if date present in filename, put it in published_date, adding time only

What happens if the existing date is not the current date?

If `cobalt publish` is setting `published_date`, do we still maintain the date extraction logic or just detect it and strip it?

**Comment by Geobert** [comment]
> What happens if the existing date is not the current date?

I think that we replace the current prefix with the current date, as `publish` is setting the `published_date` field even if it's already present.

**Comment by epage** [comment]
> I think that we replace the current prefix with the current date, as publish is setting the published_date field even if it's already present.

Updated the issue with this

**Comment by epage** [comment]
I've updated the RFC with my proposal on inferring published_date from filename.  It is based on my recent experience of putting my meetup talks on my site.  The ideal is that I have a `talks` collection that parallels my `blog` collection.  Since we can't do that ideal yet, I'm part-way doing things by having dates on regular pages.

See https://github.com/epage/epage.github.io/tree/source/blog/talks

**Comment by Geobert** [comment]
Just to be sure, the flag to activate is still `publish_date_in_filename` and not `add-date-prefix` as stated in the commit message? I've checked master, and it seems to be `publish_date_in_filename`.

**Comment by Geobert** [comment]
Tried to use it, for some reason my file is not considered in the `posts` collection. So it does not trigger the prefixing.

My blog is like this:
```
blog
  |
  -- src
        |
        -- posts
              |
              -- my_post.md
``
I'm in `blog` and call `cobalt publish ./src/posts/my_post.md`
The file get the published status and the `published_date` set, but no renaming. Adding some traces shows that `my_post.md` is not inside the collection :-/

What am I missing?

**Comment by epage** [comment]
> Just to be sure, the flag to activate is still publish_date_in_filename and not add-date-prefix as stated in the commit message? I've checked master, and it seems to be publish_date_in_filename.

I didn't pay attention when rewriting the commit message, I just took from the [existing commit message](https://github.com/cobalt-org/cobalt.rs/pull/571/commits).

**Comment by epage** [comment]
> EDIT2: fixed with #587

Thanks! Will go and look at it.
